elkscmd/basic/host-ibmpc.c

```diff
@@ -0,0 +1,359 @@
+/*
+ * Architecture Specific routines for CGA
+ * Sep 2024 Takahiro Yamada
+ */
+#include <stdio.h>
+#include <stdlib.h>
+#if __ia16__
+#include <arch/io.h>
+#else
+#define inb(port)           0
+#define inw(port)           0
+#define outb(value,port)
+#define outw(value,port)
+#endif
+#include "host.h"
+#include "basic.h"
+
+#define VIDEO_05_G320x200   0x0005
+#define VIDEO_03_T80x25     0x0003
+
+extern FILE *outfile;
+
+static int gmode = 0;
+static int exit_on = 0;
+
+typedef struct {
+    int x;
+    int y;
+    int fgc;
+    int bgc;
+    int r;
+} xyc_t;
+
+static xyc_t gxyc = {0, 0, 7, 0, 1};
+
+#if __ia16__
+void int_10(unsigned int ax, unsigned int bx,
+            unsigned int cx, unsigned int dx)
+{
+    __asm__ volatile ("push %ds;"
+                      "push %es;"
+                      "push %bp;"
+                      "push %si;"
+                      "push %di;");
+    __asm__ volatile ("int $0x10;"
```

Comment:
Have you looked at the code generated for this function using `ia16-elf-objdump -D -r -Mi8086 host-ibmpc.o`? 

The first `__asm__` directive is using a very non-standard register save sequence, and I would like to know that gcc isn't also saving the same registers on its own, which it usually does when indicating the loads required on line 47.

Also, saving BP is almost always the job of the compiler, as it will generate code using BP indirection to reference the parameters passed to the function. So I believe it should not be saved explicitly, but you need to look at the code to confirm.

I also believe that SI and DI are automatically saved by both the compiler if needed, as well as the BIOS.

After confirming which registers need to be saved, I then am wondering whether/why DS and ES need to be saved: it is my understanding that INT 10h saves segment register and likely SI and DI as well. Have you determined that these registers need to the saved in order for the code to work?

Finally, the use of a non-standard register save sequence at function prologue breaks our stack backtrace library found in libc/debug/stacktrace.c. The standard register save sequence used by ia16-elf-gcc is to save (and only when needed) SI, DI, BP (then SP->BP). After that, DS and/or ES are pushed, but only when needed. This is a complicated subject but we can talk more about it if you need more information.

Reply:
I have looked the objdump before and yes, I knew bp is saved twice,
but I haven't fixed that yet.
I don't know much about INT 10h and whether it saves ES, DS and others or not,
but maybe you are right.

I will try to write separate .S as you mentioned.

Reply:
> I don't know much about INT 10h and whether it saves ES, DS and others or not,

[Here's a reference site I use a lot](https://stanislavs.org/helppc/int_10.html). They do mention that SI and DI may get trashed on PS/2 so I would save them, but not bother with the segment registers, as unless needed, it will just serve to slow down the graphics that much more.